File: tools/muslctype.py
"""Compile the vendored musl block out of a source file and hold it to libc.

Usage: python3 tools/muslctype.py --verify <file.c>

Zero phase 15 replaces eleven libc calls and six `<ctype.h>` macros with static
functions of its own.  The claim is not "these look like musl's" but "these
compute what musl computes", and the only way to say that is to run both.  So
this slices the block OUT OF THE SOURCE THE PHASE PRODUCED -- not out of
tools/musl-ctype.txt, which would only prove the copy was faithful -- wraps it in
a main() that calls libc beside it, compiles with -Wall -Wextra, and runs it.

THE DOMAIN IS BOUNDED ON PURPOSE, and this is the one number in the phase that
is not exhaustive.  All eleven classifier/mapper functions were compared with
musl's over ALL 4,294,967,296 int values while the phase was being written:
**zero disagreements**, in 66 seconds.  Sixty-six seconds is more than twice what
the rest of the phase costs, and a check that doubles a phase's time to re-prove
a closed-form identity is a check that stops being run.  What runs here instead
is every int in [-1024, 1024], every threshold in the definitions and its two
neighbours, EOF, INT_MIN, INT_MAX, and a FIXED pseudo-random sample of 1,000,000
values spread over the whole int range -- fixed, so the check is a function of
its input like everything else here.  Every one of the eleven is a closed form in
`(unsigned)c`, so a disagreement anywhere is a disagreement at a threshold, and
the thresholds are all in the bounded set.

The other three are exhaustive over their real domains and cost nothing:

  atoi/atol   every string of length 1..4 over " \\t\\n\\r\\v\\f+-09135abx." and the
              two high bytes \\x80 \\xff -- 137,560 of them.  The high bytes are
              the sign-extension trap: they are negative as `char`, and both
              musl's macros and these answer by unsigned wrap.
  bsearch     every array size 0..40 and every key from below the first element
              to above the last, 1,845 lookups, compared BY POINTER -- which is
              stronger than "it found something", and is what makes the two
              equally wrong on an unsorted table instead of one of them quietly
              right.
  qsort       2,000 random arrays of 1..64 distinct strings, plus one array with
              three EQUAL keys.  The equal-key case is required to DIFFER: musl's
              smoothsort is unstable and this insertion sort is stable, and a
              harness that cannot tell them apart would pass a phase that had
              swapped one for the other by accident.  `sort_strings()`'s one
              caller cannot produce equal keys (`uh_seq` is `++b_u_seq_last`), so
              the difference is unobservable in the editor -- which is a fact
              about the call site and has to be proved somewhere the call site
              cannot hide it.
"""
import os
import re
import subprocess
import sys
import tempfile

START = '    static int\nmusl_isdigit(int c)\n'
END = 'static int musl_towupper(int a);\n'

MAIN = r'''
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <limits.h>

static int cmpstr(const void *a, const void *b)
{
    return strcmp(*(char *const *)a, *(char *const *)b);
}

static int cmpint(const void *a, const void *b)
{
    int x = *(const int *)a;
    int y = *(const int *)b;
    return x < y ? -1 : x > y;
}

static long bad;
static long checked;

static void one(int i)
{
    checked++;
    bad += !!musl_isalnum(i) != !!isalnum(i);
    bad += !!musl_iscntrl(i) != !!iscntrl(i);
    bad += !!musl_ispunct(i) != !!ispunct(i);
    bad += musl_tolower(i) != tolower(i);
    bad += musl_toupper(i) != toupper(i);
    bad += !!musl_isalpha(i) != !!isalpha(i);
    bad += !!musl_isdigit(i) != !!isdigit(i);
    bad += !!musl_isupper(i) != !!isupper(i);
    bad += !!musl_islower(i) != !!islower(i);
    bad += !!musl_isgraph(i) != !!isgraph(i);
    bad += !!musl_isspace(i) != !!isspace(i);
}

int main(void)
{
    static const int edge[] = {
        0x00, 0x08, 0x09, 0x0d, 0x0e, 0x1f, 0x20, 0x21, 0x2f, 0x30, 0x39, 0x3a,
        0x40, 0x41, 0x5a, 0x5b, 0x60, 0x61, 0x7a, 0x7b, 0x7e, 0x7f, 0x80, 0xfe,
        0xff, 0x100, EOF, INT_MIN, INT_MIN + 1, INT_MAX, INT_MAX - 1
    };
    unsigned seed = 2166136261u;
    int i;
    unsigned k;

    for (i = -1024; i <= 1024; i++)
    {
        one(i);
    }
    for (k = 0; k < sizeof(edge) / sizeof(edge[0]); k++)
    {
        one(edge[k]);
        if (edge[k] != INT_MAX)
        {
            one(edge[k] + 1);
        }
        if (edge[k] != INT_MIN)
        {
            one(edge[k] - 1);
        }
    }
    for (k = 0; k < 1000000u; k++)
    {
        seed = seed * 1103515245u + 12345u;
        one((int)seed);
    }
    printf("ctype %ld %ld\n", checked, bad);

    {
        static const char al[] = " \t\n\r\v\f+-09135abx.\x80\xff";
        int A = (int)sizeof(al) - 1;
        long cnt = 0;
        long bada = 0;
        int len;
        char buf[8];

        for (len = 1; len <= 4; len++)
        {
            long tot = 1;
            long t;
            int p;

            for (p = 0; p < len; p++)
            {
                tot *= A;
            }
            for (t = 0; t < tot; t++)
            {
                long v = t;

                for (p = 0; p < len; p++)
                {
                    buf[p] = al[v % A];
                    v /= A;
                }
                buf[len] = 0;
                cnt++;
                bada += musl_atoi(buf) != atoi(buf);
                bada += musl_atol(buf) != atol(buf);
            }
        }
        printf("atoi %ld %ld\n", cnt, bada);
    }

    {
        int arr[41];
        int sz;
        int key;
        long cnt = 0;
        long badb = 0;

        for (sz = 0; sz <= 40; sz++)
        {
            for (i = 0; i < sz; i++)
            {
                arr[i] = i * 2;
            }
            for (key = -2; key <= sz * 2 + 2; key++)
            {
                void *a = bsearch(&key, arr, (size_t)sz, sizeof(int), cmpint);
                void *b = musl_bsearch(&key, arr, (size_t)sz, sizeof(int), cmpint);

                cnt++;
                badb += a != b;
            }
        }
        printf("bsearch %ld %ld\n", cnt, badb);
    }

    {
        char *p[64];
        char *q[64];
        static char store[64][8];
        long badq = 0;
        int t;
        int sz;

        seed = 12345u;
        for (t = 0; t < 2000; t++)
        {
            sz = (int)(seed % 64u) + 1;
            seed = seed * 1103515245u + 12345u;
            for (i = 0; i < sz; i++)
            {
                snprintf(store[i], sizeof store[i], "%06u", seed % 1000000u);
                seed = seed * 1103515245u + 12345u;
                p[i] = store[i];
                q[i] = store[i];
            }
            qsort(p, (size_t)sz, sizeof p[0], cmpstr);
            musl_qsort(q, (size_t)sz, sizeof q[0], cmpstr);
            for (i = 0; i < sz; i++)
            {
                if (strcmp(p[i], q[i]) != 0)
                {
                    badq++;
                    break;
                }
            }
        }
        printf("qsort 2000 %ld\n", badq);
    }

    {
        static char a[] = "aa";
        static char b1[] = "bb";
        static char b2[] = "bb";
        static char b3[] = "bb";
        static char c[] = "cc";
        char *src[5];
        char *p[5];
        char *q[5];
        int moved = 0;
        int order = 0;

        src[0] = b1; src[1] = c; src[2] = b2; src[3] = a; src[4] = b3;
        for (i = 0; i < 5; i++)
        {
            p[i] = src[i];
            q[i] = src[i];
        }
        qsort(p, 5, sizeof p[0], cmpstr);
        musl_qsort(q, 5, sizeof q[0], cmpstr);
        for (i = 0; i < 5; i++)
        {
            moved += p[i] != q[i];
            order += strcmp(p[i], q[i]) != 0;
        }
        printf("ties %d %d\n", moved, order);
    }
    return 0;
}
'''

TAG = 'vendor'


def die(msg):
    sys.exit('  %-12s %s' % (TAG, msg))


def main():
    if len(sys.argv) != 3 or sys.argv[1] != '--verify':
        sys.exit('usage: muslctype.py --verify <file.c>')
    path = sys.argv[2]
    text = open(path, errors='surrogateescape').read()
    if text.count(START) != 1 or text.count(END) != 1:
        die('%s does not hold exactly one vendored block' % path)
    a = text.index(START)
    z = text.index(END)
    if z <= a:
        die('the vendored block and its two prototypes are in the wrong order')
    block = text[a:z]
    for name in ('musl_isdigit', 'musl_isalpha', 'musl_isupper', 'musl_islower',
                 'musl_isgraph', 'musl_isspace', 'musl_isalnum', 'musl_iscntrl',
                 'musl_ispunct', 'musl_tolower', 'musl_toupper', 'musl_atoi',
                 'musl_atol', 'musl_bsearch', 'musl_qsort'):
        if len(re.findall(r'^%s\(' % name, block, re.M)) != 1:
            die('the block sliced out of %s does not define %s exactly once'
                % (path, name))

    with tempfile.TemporaryDirectory() as d:
        src = os.path.join(d, 'h.c')
        open(src, 'w').write('#include <stddef.h>\n' + block + MAIN)
        r = subprocess.run(['gcc', '-O2', '-Wall', '-Wextra', '-o',
                            os.path.join(d, 'h'), src],
                           capture_output=True, text=True)
        if r.returncode or r.stderr.strip():
            print('  %-12s the vendored block does not compile clean on its own:' % TAG)
            print(r.stderr.rstrip()[:2000])
            sys.exit(1)
        r = subprocess.run([os.path.join(d, 'h')], capture_output=True, text=True)
        if r.returncode:
            die('the harness did not run')
        got = dict(line.split(None, 1) for line in r.stdout.splitlines())

    def num(key, idx):
        return int(got[key].split()[idx])

    fail = []
    if num('ctype', 1) != 0:
        fail.append('%d ctype disagreements with libc' % num('ctype', 1))
    if num('ctype', 0) < 1002000:
        fail.append('the ctype domain shrank to %d values' % num('ctype', 0))
    if num('atoi', 0) != 137560 or num('atoi', 1) != 0:
        fail.append('atoi/atol: %s' % got['atoi'])
    if num('bsearch', 0) != 1845 or num('bsearch', 1) != 0:
        fail.append('bsearch: %s' % got['bsearch'])
    if num('qsort', 1) != 0:
        fail.append('qsort: %s' % got['qsort'])
    if num('ties', 1) != 0:
        fail.append('the tie case sorts to a DIFFERENT string order, which is a bug '
                    'in musl_qsort and not a tie-break: %s' % got['ties'])
    if num('ties', 0) == 0:
        fail.append('the tie case places every pointer exactly as musl does, so this '
                    'harness cannot tell a stable sort from an unstable one and the '
                    'qsort result above proves nothing')
    if fail:
        for line in fail:
            print('  %-12s %s' % (TAG, line))
        sys.exit(1)
    print('  %-12s the block compiled OUT OF THE PRODUCED SOURCE agrees with libc: '
          '%d int values (bounded -- all 2^32 were checked once, 0 disagreements, and '
          'it costs 66 s), %d atoi/atol strings, %d bsearch lookups compared by '
          'POINTER, 2000 qsort arrays -- and the three-equal-keys case moves %d of 5 '
          'pointers while sorting to the same strings, which is what says the harness '
          'can see a tie at all'
          % (TAG, num('ctype', 0), num('atoi', 0), num('bsearch', 0), num('ties', 0)))


if __name__ == '__main__':
    main()

File: tools/go/internal/harness/muslctype.go
package harness

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
)

// MuslCtypeVerify is tools/muslctype.py: compile the vendored musl block OUT OF
// THE SOURCE THE PHASE PRODUCED and hold it to the libc this machine links.
//
// The claim zero phase 15 makes is not "these look like musl's" but "these
// compute what musl computes", and the only way to say that is to run both.  So
// the block is sliced out of the .c the phase wrote -- not out of
// tools/musl-ctype.txt, which would only prove the copy was faithful -- wrapped
// in a main() that calls libc beside it, compiled with -Wall -Wextra, and run.
//
// THE DOMAIN IS BOUNDED ON PURPOSE and it is the one number here that is not
// exhaustive.  All eleven classifiers were compared with musl's over all
// 4,294,967,296 int values while the phase was written: zero disagreements, in
// 66 seconds.  A check that doubles a phase's time to re-prove a closed form is
// a check that stops being run.  What runs is every int in [-1024, 1024], every
// threshold and its two neighbours, EOF, INT_MIN, INT_MAX, and a FIXED
// pseudo-random sample of 1,000,000 -- fixed, so the check is a function of its
// input like everything else here.
//
// THE C DRIVER BELOW IS GENERATED from tools/muslctype.py by
// tools/gocmp/genmuslctype.py, which imports that module rather than scraping
// it.  Do not hand-edit it: a changed constant in the random walk or a dropped
// row in the edge table would be invisible to both the compile and the
// comparison.
const muslCtypeStart = "    static int\n" +
	"musl_isdigit(int c)\n" +
	""

const muslCtypeEnd = "static int musl_towupper(int a);\n" +
	""

const muslCtypeMain = "\n" +
	"#include <stdio.h>\n" +
	"#include <stdlib.h>\n" +
	"#include <string.h>\n" +
	"#include <ctype.h>\n" +
	"#include <limits.h>\n" +
	"\n" +
	"static int cmpstr(const void *a, const void *b)\n" +
	"{\n" +
	"    return strcmp(*(char *const *)a, *(char *const *)b);\n" +
	"}\n" +
	"\n" +
	"static int cmpint(const void *a, const void *b)\n" +
	"{\n" +
	"    int x = *(const int *)a;\n" +
	"    int y = *(const int *)b;\n" +
	"    return x < y ? -1 : x > y;\n" +
	"}\n" +
	"\n" +
	"static long bad;\n" +
	"static long checked;\n" +
	"\n" +
	"static void one(int i)\n" +
	"{\n" +
	"    checked++;\n" +
	"    bad += !!musl_isalnum(i) != !!isalnum(i);\n" +
	"    bad += !!musl_iscntrl(i) != !!iscntrl(i);\n" +
	"    bad += !!musl_ispunct(i) != !!ispunct(i);\n" +
	"    bad += musl_tolower(i) != tolower(i);\n" +
	"    bad += musl_toupper(i) != toupper(i);\n" +
	"    bad += !!musl_isalpha(i) != !!isalpha(i);\n" +
	"    bad += !!musl_isdigit(i) != !!isdigit(i);\n" +
	"    bad += !!musl_isupper(i) != !!isupper(i);\n" +
	"    bad += !!musl_islower(i) != !!islower(i);\n" +
	"    bad += !!musl_isgraph(i) != !!isgraph(i);\n" +
	"    bad += !!musl_isspace(i) != !!isspace(i);\n" +
	"}\n" +
	"\n" +
	"int main(void)\n" +
	"{\n" +
	"    static const int edge[] = {\n" +
	"        0x00, 0x08, 0x09, 0x0d, 0x0e, 0x1f, 0x20, 0x21, 0x2f, 0x30, 0x39, 0x3a,\n" +
	"        0x40, 0x41, 0x5a, 0x5b, 0x60, 0x61, 0x7a, 0x7b, 0x7e, 0x7f, 0x80, 0xfe,\n" +
	"        0xff, 0x100, EOF, INT_MIN, INT_MIN + 1, INT_MAX, INT_MAX - 1\n" +
	"    };\n" +
	"    unsigned seed = 2166136261u;\n" +
	"    int i;\n" +
	"    unsigned k;\n" +
	"\n" +
	"    for (i = -1024; i <= 1024; i++)\n" +
	"    {\n" +
	"        one(i);\n" +
	"    }\n" +
	"    for (k = 0; k < sizeof(edge) / sizeof(edge[0]); k++)\n" +
	"    {\n" +
	"        one(edge[k]);\n" +
	"        if (edge[k] != INT_MAX)\n" +
	"        {\n" +
	"            one(edge[k] + 1);\n" +
	"        }\n" +
	"        if (edge[k] != INT_MIN)\n" +
	"        {\n" +
	"            one(edge[k] - 1);\n" +
	"        }\n" +
	"    }\n" +
	"    for (k = 0; k < 1000000u; k++)\n" +
	"    {\n" +
	"        seed = seed * 1103515245u + 12345u;\n" +
	"        one((int)seed);\n" +
	"    }\n" +
	"    printf(\"ctype %ld %ld\\n\", checked, bad);\n" +
	"\n" +
	"    {\n" +
	"        static const char al[] = \" \\t\\n\\r\\v\\f+-09135abx.\\x80\\xff\";\n" +
	"        int A = (int)sizeof(al) - 1;\n" +
	"        long cnt = 0;\n" +
	"        long bada = 0;\n" +
	"        int len;\n" +
	"        char buf[8];\n" +
	"\n" +
	"        for (len = 1; len <= 4; len++)\n" +
	"        {\n" +
	"            long tot = 1;\n" +
	"            long t;\n" +
	"            int p;\n" +
	"\n" +
	"            for (p = 0; p < len; p++)\n" +
	"            {\n" +
	"                tot *= A;\n" +
	"            }\n" +
	"            for (t = 0; t < tot; t++)\n" +
	"            {\n" +
	"                long v = t;\n" +
	"\n" +
	"                for (p = 0; p < len; p++)\n" +
	"                {\n" +
	"                    buf[p] = al[v % A];\n" +
	"                    v /= A;\n" +
	"                }\n" +
	"                buf[len] = 0;\n" +
	"                cnt++;\n" +
	"                bada += musl_atoi(buf) != atoi(buf);\n" +
	"                bada += musl_atol(buf) != atol(buf);\n" +
	"            }\n" +
	"        }\n" +
	"        printf(\"atoi %ld %ld\\n\", cnt, bada);\n" +
	"    }\n" +
	"\n" +
	"    {\n" +
	"        int arr[41];\n" +
	"        int sz;\n" +
	"        int key;\n" +
	"        long cnt = 0;\n" +
	"        long badb = 0;\n" +
	"\n" +
	"        for (sz = 0; sz <= 40; sz++)\n" +
	"        {\n" +
	"            for (i = 0; i < sz; i++)\n" +
	"            {\n" +
	"                arr[i] = i * 2;\n" +
	"            }\n" +
	"            for (key = -2; key <= sz * 2 + 2; key++)\n" +
	"            {\n" +
	"                void *a = bsearch(&key, arr, (size_t)sz, sizeof(int), cmpint);\n" +
	"                void *b = musl_bsearch(&key, arr, (size_t)sz, sizeof(int), cmpint);\n" +
	"\n" +
	"                cnt++;\n" +
	"                badb += a != b;\n" +
	"            }\n" +
	"        }\n" +
	"        printf(\"bsearch %ld %ld\\n\", cnt, badb);\n" +
	"    }\n" +
	"\n" +
	"    {\n" +
	"        char *p[64];\n" +
	"        char *q[64];\n" +
	"        static char store[64][8];\n" +
	"        long badq = 0;\n" +
	"        int t;\n" +
	"        int sz;\n" +
	"\n" +
	"        seed = 12345u;\n" +
	"        for (t = 0; t < 2000; t++)\n" +
	"        {\n" +
	"            sz = (int)(seed % 64u) + 1;\n" +
	"            seed = seed * 1103515245u + 12345u;\n" +
	"            for (i = 0; i < sz; i++)\n" +
	"            {\n" +
	"                snprintf(store[i], sizeof store[i], \"%06u\", seed % 1000000u);\n" +
	"                seed = seed * 1103515245u + 12345u;\n" +
	"                p[i] = store[i];\n" +
	"                q[i] = store[i];\n" +
	"            }\n" +
	"            qsort(p, (size_t)sz, sizeof p[0], cmpstr);\n" +
	"            musl_qsort(q, (size_t)sz, sizeof q[0], cmpstr);\n" +
	"            for (i = 0; i < sz; i++)\n" +
	"            {\n" +
	"                if (strcmp(p[i], q[i]) != 0)\n" +
	"                {\n" +
	"                    badq++;\n" +
	"                    break;\n" +
	"                }\n" +
	"            }\n" +
	"        }\n" +
	"        printf(\"qsort 2000 %ld\\n\", badq);\n" +
	"    }\n" +
	"\n" +
	"    {\n" +
	"        static char a[] = \"aa\";\n" +
	"        static char b1[] = \"bb\";\n" +
	"        static char b2[] = \"bb\";\n" +
	"        static char b3[] = \"bb\";\n" +
	"        static char c[] = \"cc\";\n" +
	"        char *src[5];\n" +
	"        char *p[5];\n" +
	"        char *q[5];\n" +
	"        int moved = 0;\n" +
	"        int order = 0;\n" +
	"\n" +
	"        src[0] = b1; src[1] = c; src[2] = b2; src[3] = a; src[4] = b3;\n" +
	"        for (i = 0; i < 5; i++)\n" +
	"        {\n" +
	"            p[i] = src[i];\n" +
	"            q[i] = src[i];\n" +
	"        }\n" +
	"        qsort(p, 5, sizeof p[0], cmpstr);\n" +
	"        musl_qsort(q, 5, sizeof q[0], cmpstr);\n" +
	"        for (i = 0; i < 5; i++)\n" +
	"        {\n" +
	"            moved += p[i] != q[i];\n" +
	"            order += strcmp(p[i], q[i]) != 0;\n" +
	"        }\n" +
	"        printf(\"ties %d %d\\n\", moved, order);\n" +
	"    }\n" +
	"    return 0;\n" +
	"}\n" +
	""

// muslCtypeNames are the fifteen the block must define exactly once each.  The
// slice is by text, so a block that lost a function would still compile if
// libc supplied the name -- this is what stops that.
var muslCtypeNames = []string{
	"musl_isdigit", "musl_isalpha", "musl_isupper", "musl_islower",
	"musl_isgraph", "musl_isspace", "musl_isalnum", "musl_iscntrl",
	"musl_ispunct", "musl_tolower", "musl_toupper", "musl_atoi",
	"musl_atol", "musl_bsearch", "musl_qsort",
}

const muslCtypeTag = "vendor"

func muslDie(format string, a ...any) error {
	return fmt.Errorf("  %-12s %s", muslCtypeTag, fmt.Sprintf(format, a...))
}

// MuslCtypeVerify returns nil when the vendored block agrees with libc, and
// writes its own success line -- which carries the numbers, because a check
// that prints only "ok" cannot be read for whether it did anything.
func MuslCtypeVerify(path string, out *os.File) error {
	data, err := os.ReadFile(path)
	if err != nil {
		return err
	}
	text := string(data)
	if strings.Count(text, muslCtypeStart) != 1 || strings.Count(text, muslCtypeEnd) != 1 {
		return muslDie("%s does not hold exactly one vendored block", path)
	}
	a := strings.Index(text, muslCtypeStart)
	z := strings.Index(text, muslCtypeEnd)
	if z <= a {
		return muslDie("the vendored block and its two prototypes are in the wrong order")
	}
	block := text[a:z]
	for _, name := range muslCtypeNames {
		re := regexp.MustCompile("(?m)^" + regexp.QuoteMeta(name) + `\(`)
		if len(re.FindAllString(block, -1)) != 1 {
			return muslDie("the block sliced out of %s does not define %s exactly once", path, name)
		}
	}

	d, err := os.MkdirTemp("", "muslctype")
	if err != nil {
		return err
	}
	defer os.RemoveAll(d)
	src := filepath.Join(d, "h.c")
	if err := os.WriteFile(src, []byte("#include <stddef.h>\n"+block+muslCtypeMain), 0644); err != nil {
		return err
	}
	bin := filepath.Join(d, "h")
	cc := exec.Command("gcc", "-O2", "-Wall", "-Wextra", "-o", bin, src)
	var ccerr strings.Builder
	cc.Stderr = &ccerr
	// A CLEAN COMPILE IS PART OF THE CLAIM, so a warning fails it: gcc exits 0
	// with warnings, and a check that tests only the status passes always.
	if err := cc.Run(); err != nil || strings.TrimSpace(ccerr.String()) != "" {
		fmt.Fprintf(out, "  %-12s the vendored block does not compile clean on its own:\n", muslCtypeTag)
		s := strings.TrimRight(ccerr.String(), "\n")
		if len(s) > 2000 {
			s = s[:2000]
		}
		fmt.Fprintln(out, s)
		return ErrReported
	}
	run := exec.Command(bin)
	stdout, err := run.Output()
	if err != nil {
		return muslDie("the harness did not run")
	}

	got := map[string][]string{}
	for _, line := range strings.Split(string(stdout), "\n") {
		f := strings.Fields(line)
		if len(f) > 1 {
			got[f[0]] = f[1:]
		}
	}
	num := func(key string, idx int) (int, error) {
		v, ok := got[key]
		if !ok || idx >= len(v) {
			return 0, muslDie("the harness printed no %s row, so its result cannot be read", key)
		}
		n, err := strconv.Atoi(v[idx])
		if err != nil {
			return 0, muslDie("the harness's %s row is not a number: %v", key, v)
		}
		return n, nil
	}

	var fail []string
	get := func(key string, idx int) int {
		n, err := num(key, idx)
		if err != nil {
			fail = append(fail, err.Error())
			return -1
		}
		return n
	}
	ctypeN, ctypeBad := get("ctype", 0), get("ctype", 1)
	atoiN, atoiBad := get("atoi", 0), get("atoi", 1)
	bsN, bsBad := get("bsearch", 0), get("bsearch", 1)
	qsBad := get("qsort", 1)
	tiesMoved, tiesOrder := get("ties", 0), get("ties", 1)

	if ctypeBad != 0 {
		fail = append(fail, fmt.Sprintf("%d ctype disagreements with libc", ctypeBad))
	}
	if ctypeN < 1002000 {
		fail = append(fail, fmt.Sprintf("the ctype domain shrank to %d values", ctypeN))
	}
	if atoiN != 137560 || atoiBad != 0 {
		fail = append(fail, fmt.Sprintf("atoi/atol: %d %d", atoiN, atoiBad))
	}
	if bsN != 1845 || bsBad != 0 {
		fail = append(fail, fmt.Sprintf("bsearch: %d %d", bsN, bsBad))
	}
	if qsBad != 0 {
		fail = append(fail, fmt.Sprintf("qsort: %v", got["qsort"]))
	}
	// THE TIE CASE IS REQUIRED TO DIFFER IN POINTERS AND NOT IN STRINGS.
	// musl's smoothsort is unstable and this insertion sort is stable, so a
	// harness that could not tell them apart would pass a phase that had
	// swapped one for the other by accident.
	if tiesOrder != 0 {
		fail = append(fail, fmt.Sprintf("the tie case sorts to a DIFFERENT string order, "+
			"which is a bug in musl_qsort and not a tie-break: %v", got["ties"]))
	}
	if tiesMoved == 0 {
		fail = append(fail, "the tie case places every pointer exactly as musl does, so this "+
			"harness cannot tell a stable sort from an unstable one and the "+
			"qsort result above proves nothing")
	}
	if len(fail) > 0 {
		for _, line := range fail {
			fmt.Fprintf(out, "  %-12s %s\n", muslCtypeTag, strings.TrimSpace(strings.TrimPrefix(line, "  "+muslCtypeTag)))
		}
		return ErrReported
	}
	fmt.Fprintf(out, "  %-12s the block compiled OUT OF THE PRODUCED SOURCE agrees with libc: "+
		"%d int values (bounded -- all 2^32 were checked once, 0 disagreements, and "+
		"it costs 66 s), %d atoi/atol strings, %d bsearch lookups compared by "+
		"POINTER, 2000 qsort arrays -- and the three-equal-keys case moves %d of 5 "+
		"pointers while sorting to the same strings, which is what says the harness "+
		"can see a tie at all\n",
		muslCtypeTag, ctypeN, atoiN, bsN, tiesMoved)
	return nil
}
